flagscale/serve/core/dag.py

```diff
@@ -0,0 +1,455 @@
+import os, importlib
+import sys
+import uvicorn
+import socket
+import subprocess
+import json
+import numpy as np
+import omegaconf
+import matplotlib.pyplot as plt
+
+from pathlib import Path
+from pydantic import create_model
+from typing import Callable, Any
+from fastapi import FastAPI, HTTPException, Request
+import ray
+from ray import workflow
+
+from flagscale.logger import logger
+
+
+class Builder:
+    def __init__(self, config):
+        self.config = config.serve
+        self.exp_config = config.experiment
+        self.check_config(self.config)
+        self.tasks = {}
+
+    def check_config(self, config):
+        if not config.get("deploy", None):
+            raise ValueError("key deploy is missing for deployment configuration.")
+        if not config.deploy.get("models", None):
+            raise ValueError("key models is missing for building dag pipeline.")
+
+    def find_final_node(self):
+        whole_nodes = set(self.config["deploy"]["models"].keys())
+        dependencies = set()
+
+        for model_alias, model_config in self.config["deploy"]["models"].items():
+            if len(model_config.get("depends", [])) > 0:
+                dependencies.update(model_config.depends)
+
+        output_node = whole_nodes - dependencies
+        if len(output_node) != 1:
+            raise ValueError(
+                f"There should only have one final node but there are {len(output_node)} nodes {output_node}."
+            )
+        return list(output_node)[0]
+
+    def check_and_get_port(self, target_port=None, host="0.0.0.0"):
+        """
+        Check if a specific port is free; if not, allocate a free port.
+        :param target_port: The port number to check, default is None.
+        :param host: The host address to check, default is "0.0.0.0".
+        :return: A tuple (is_free, port), where `is_free` indicates if the target port is free,
+                and `port` is the allocated port (target_port if free, or a new free port).
+        """
```

Comment:
The code should be placed in the utilities so that it can be reused.

Reply:
ok